File: 백준/Silver/1966. 프린터 큐/프린터 큐.java
import java.io.*;
import java.util.*;
import java.util.stream.Stream;

public class Main {

    public static void main(String[] args) throws Exception {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int tc = Integer.parseInt(br.readLine());

        for (int i=0; i<tc; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int N = Integer.parseInt(st.nextToken());
            int M = Integer.parseInt(st.nextToken());

            Queue<int[]> q = new LinkedList<int[]>();

            st = new StringTokenizer(br.readLine());

            for (int j=0; j<N; j++) {
                int num = Integer.parseInt(st.nextToken());
                q.add(new int[] {j, num});
            }

            int cnt = 0;

            while (true) {
                int[] pollQ = q.poll();
                boolean check = true;

                for (int[] num: q){
                    if (num[1] > pollQ[1]) {
                        check = false;
                        break;
                    }
                }

                if (check) {
                    cnt++;
                    if (pollQ[0] == M) break;
                } else {
                    q.add(pollQ);
                }
            }

            System.out.println(cnt);
        }
    }
}

File: 백준/Silver/1966. 프린터 큐/프린터 큐.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*

fun main() {
    /*
     * T: 테스트케이스
     * N: 문서 개수
     * M: 궁금한 문서 순서 (맨 왼쪽은 0)
     * N개의 문서 중요도가 차례로 주어짐
     **/

    val br = BufferedReader(InputStreamReader(System.`in`))
    val T = br.readLine().toInt()

    repeat(T) {
        val (N, M) = br.readLine().split(" ").map { it.toInt() }
        val arr = br.readLine().split(" ").map { it.toInt() }

        println(findDoc(arr, M))
    }
}

fun findDoc(arr: List<Int>, M:Int):Int {
    val queue = LinkedList<Pair<Int, Int>>()
    val pq = PriorityQueue<Int>(reverseOrder())

    for (i in arr.indices) {
        queue.offer(Pair(arr[i],i))
        pq.offer(arr[i])
    }

    var printOrder = 0
    while (queue.isNotEmpty()) {
        val current = queue.peek()
        val currentPriority = current.first
        val currentIndex = current.second

        val maxPriority = pq.peek()

        if (currentPriority == maxPriority) {
            queue.poll()
            pq.poll()
            printOrder++

            if (currentIndex == M) {
                return printOrder
            }
        } else {
            queue.poll()
            queue.offer(current)
        }
    }

    return -1
}
